### PR TITLE
quic: refactor use preferred address strategy

### PR DESCRIPTION
**Pull request by jasnell**
Refactor preferred address strategy internals + add usePreferredAddress
event on the JavaScript so user code can know when the remote address
has been modified.

### CODE REVIEW

File: doc/api/quic.md [hunk 1]
```diff
@@ -847,6 +847,23 @@ three arguments:
 The `sessionTicket` and `remoteTransportParams` are useful when creating a new
 `QuicClientSession` to more quickly resume an existing session.
 
+#### Event: `'usePreferredAddress'`
+<!-- YAML
+added: REPLACEME
+-->
+
+The `'usePreferredAddress'` event is emitted when the client `QuicSession`
+is updated to use the server advertised preferred address.The callback is
```
Comment (addaleax) [inline doc/api/quic.md]:
```suggestion
is updated to use the server-advertised preferred address. The callback is
```

File: doc/api/quic.md [hunk 2]
```diff
@@ -847,6 +847,23 @@ three arguments:
...
+invoked with a single `address` argument:
+
+* `address` {Object}
+  * `address` {string} The preferred hostname
+  * `port` {number} The preferred IP port
+  * `type` {string} Either 'udp4' or 'udp6'.
```
Comment (addaleax) [inline doc/api/quic.md]:
```suggestion
  * `type` {string} Either `'udp4'` or `'udp6'`.
```

File: src/quic/node_quic_util-inl.h [hunk 3]
```diff
@@ -196,11 +196,21 @@ const uint8_t* QuicPreferredAddress::StatelessResetToken() const {
...
+  char host[NI_MAXHOST];
+  // Return an empty string if unable to convert...
+  if (uv_inet_ntop(AF_INET6, paddr_->ipv6_addr, host, sizeof(host)) != 0)
+    return std::string();
+
+  return std::string(host);
 }
-const sockaddr_in* QuicPreferredAddress::PreferredIPv4Address() const {
-  return reinterpret_cast<const sockaddr_in*>(&paddr_->ipv4_addr);
+const std::string QuicPreferredAddress::PreferredIPv4Address() const {
```
Comment (addaleax) [inline src/quic/node_quic_util-inl.h]: Adding `const` to return types only leads to compiler warnings :)

```suggestion
std::string QuicPreferredAddress::PreferredIPv4Address() const {
```

File: src/quic/node_quic_util-inl.h [hunk 4]
```diff
@@ -196,11 +196,21 @@ const uint8_t* QuicPreferredAddress::StatelessResetToken() const {
   return paddr_->stateless_reset_token;
 }
 
-const sockaddr_in6* QuicPreferredAddress::PreferredIPv6Address() const {
-  return reinterpret_cast<const sockaddr_in6*>(&paddr_->ipv6_addr);
+const std::string QuicPreferredAddress::PreferredIPv6Address() const {
```
Comment (addaleax) [inline src/quic/node_quic_util-inl.h]:
```suggestion
std::string QuicPreferredAddress::PreferredIPv6Address() const {
```

File: src/quic/node_quic_util.h [hunk 5]
```diff
@@ -88,8 +88,8 @@ class QuicPreferredAddress {
       paddr_(paddr) {}
 
   inline const ngtcp2_cid* cid() const;
-  inline const sockaddr_in6* PreferredIPv6Address() const;
-  inline const sockaddr_in* PreferredIPv4Address() const;
+  inline const std::string PreferredIPv6Address() const;
```
Comment (addaleax) [inline src/quic/node_quic_util.h]:
```suggestion
  inline std::string PreferredIPv6Address() const;
```

File: src/quic/node_quic_util.h [hunk 6]
```diff
@@ -88,8 +88,8 @@ class QuicPreferredAddress {
...
+  inline const std::string PreferredIPv4Address() const;
```
Comment (addaleax) [inline src/quic/node_quic_util.h]:
```suggestion
  inline std::string PreferredIPv4Address() const;
```

**Comment by jasnell** [comment]
This needs a quick rebase after landing https://github.com/nodejs/quic/pull/252

**Comment by jasnell** [comment]
Landed

**Pull request closed without merging**